File: 2025-09-25 - ListaPrioridade/Exemplo_Java/ListaPrioridade.java
// ------------------------------------------------------------------------
// ListaPrioridade.java
// Implementação da Lista de Prioridade baseada no semestre
// ------------------------------------------------------------------------
public class ListaPrioridade implements ListaPrioridadeInterface {

   private Node inicio;   // Primeiro nó da lista
   private int tamanho;   // Controle do tamanho da lista

   // ---------------------------------------------------------------------
   // Construtor
   // ---------------------------------------------------------------------
   public ListaPrioridade() {
      this.inicio = null;
      this.tamanho = 0;
   }

   // ---------------------------------------------------------------------
   // Inserir nó mantendo a prioridade (menor semestre = maior prioridade)
   // ---------------------------------------------------------------------
   @Override
   public boolean inserir(Node novoNode) {
      if (novoNode == null) return false;

      // Evitar duplicidade de código
      if (buscar(novoNode.getCodigoDisciplina()) != null) return false;

      // Inserção no início se lista vazia ou prioridade maior
      if (inicio == null || novoNode.getSemestre() < inicio.getSemestre()) {
         novoNode.setProximo(inicio);
         inicio = novoNode;
      } else {
         Node atual = inicio;
         while (atual.getProximo() != null && atual.getProximo().getSemestre() <= novoNode.getSemestre()) {
            atual = atual.getProximo();
         }
         novoNode.setProximo(atual.getProximo());
         atual.setProximo(novoNode);
      }

      tamanho++;
      return true;
   }

   // ---------------------------------------------------------------------
   // Remover nó pelo código da disciplina
   // ---------------------------------------------------------------------
   @Override
   public boolean remover(String codigoDisciplina) {
      if (inicio == null || codigoDisciplina == null) return false;

      if (inicio.getCodigoDisciplina().equals(codigoDisciplina)) {
         inicio = inicio.getProximo();
         tamanho--;
         return true;
      }

      Node atual = inicio;
      while (atual.getProximo() != null && !atual.getProximo().getCodigoDisciplina().equals(codigoDisciplina)) {
         atual = atual.getProximo();
      }

      if (atual.getProximo() != null) {
         atual.setProximo(atual.getProximo().getProximo());
         tamanho--;
         return true;
      }

      return false; // Não encontrado
   }

   // ---------------------------------------------------------------------
   // Buscar nó pelo código da disciplina
   // ---------------------------------------------------------------------
   @Override
   public Node buscar(String codigoDisciplina) {
      Node atual = inicio;
      while (atual != null) {
         if (atual.getCodigoDisciplina().equals(codigoDisciplina)) return atual;
         atual = atual.getProximo();
      }
      return null;
   }

   // ---------------------------------------------------------------------
   // Atualizar disciplina existente
   // ---------------------------------------------------------------------
   @Override
   public boolean atualizar(Node nodeAtualizado) {
      if (nodeAtualizado == null) return false;

      Node atual = inicio;
      Node anterior = null;
      while (atual != null) {
         if (atual.getCodigoDisciplina().equals(nodeAtualizado.getCodigoDisciplina())) {
            // Remover o nó atual
            if (anterior == null) {
               inicio = atual.getProximo();
            } else {
               anterior.setProximo(atual.getProximo());
            }
            tamanho--;
            // Reinsere para manter prioridade
            return inserir(nodeAtualizado);
         }
         anterior = atual;
         atual = atual.getProximo();
      }

      return false; // Não encontrado
   }

   // ---------------------------------------------------------------------
   // Imprimir lista completa
   // ---------------------------------------------------------------------
   @Override
   public void imprimirLista() {
      Node atual = inicio;
      System.out.println("Lista de Prioridade:");
      while (atual != null) {
         System.out.println(atual.getCodigoDisciplina() + " – " + atual.getNomeDisciplina() +
            " (" + atual.getCargaHoraria() + "h) - Semestre " + atual.getSemestre());
         atual = atual.getProximo();
      }
   }

   // ---------------------------------------------------------------------
   // Retornar tamanho da lista
   // ---------------------------------------------------------------------
   @Override
   public int tamanho() {
      return tamanho;
   }
}

File: 2025-09-25 - ListaPrioridade/Exemplo_Python/ListaPrioridade.py
# ------------------------------------------------------------------------
# ListaPrioridade.py
# Implementação da Lista de Prioridade baseada no semestre
# ------------------------------------------------------------------------
from Node import Node
from ListaPrioridadeInterface import ListaPrioridadeInterface

class ListaPrioridade(ListaPrioridadeInterface):
    def __init__(self):
        self._inicio = None
        self._tamanhoLista = 0

    def inserir(self, novoNode: Node) -> bool:
        """ Inserir nó mantendo a prioridade (menor semestre = maior prioridade) """
        if not novoNode:
            return False
        
        if self.buscar(novoNode.getCodigoDisciplina()):
            return False

        if self._inicio is None or novoNode.getSemestre() < self._inicio.getSemestre():
            novoNode.setProximo(self._inicio)
            self._inicio = novoNode
        else:
            atual = self._inicio
            while (atual.getProximo() is not None and 
                   atual.getProximo().getSemestre() <= novoNode.getSemestre()):
                atual = atual.getProximo()
            novoNode.setProximo(atual.getProximo())
            atual.setProximo(novoNode)
        
        self._tamanhoLista += 1
        return True

    def remover(self, codigoDisciplina: str) -> bool:
        """ Remover nó pelo código da disciplina """
        if self._inicio is None or codigoDisciplina is None:
            return False

        if self._inicio.getCodigoDisciplina() == codigoDisciplina:
            self._inicio = self._inicio.getProximo()
            self._tamanhoLista -= 1
            return True

        atual = self._inicio
        while (atual.getProximo() is not None and 
               atual.getProximo().getCodigoDisciplina() != codigoDisciplina):
            atual = atual.getProximo()

        if atual.getProximo() is not None:
            proximoNode = atual.getProximo().getProximo()
            atual.setProximo(proximoNode)
            self._tamanhoLista -= 1
            return True

        return False

    def buscar(self, codigoDisciplina: str) -> Node:
        """ Buscar nó pelo código da disciplina """
        atual = self._inicio
        while atual is not None:
            if atual.getCodigoDisciplina() == codigoDisciplina:
                return atual
            atual = atual.getProximo()
        return None

    def atualizar(self, nodeAtualizado: Node) -> bool:
        """ Atualizar disciplina existente """
        if not nodeAtualizado:
            return False

        if self.remover(nodeAtualizado.getCodigoDisciplina()):
            return self.inserir(nodeAtualizado)

        return False

    def imprimirLista(self) -> None:
        """ Imprimir lista completa """
        atual = self._inicio
        print("Lista de Prioridade:")
        while atual is not None:
            print(f"{atual.getCodigoDisciplina()} – {atual.getNomeDisciplina()} "
                  f"({atual.getCargaHoraria()}h) - Semestre {atual.getSemestre()}")
            atual = atual.getProximo()

    def tamanho(self) -> int:
        """ Retornar tamanho da lista """
        return self._tamanhoLista